### PR TITLE
feat: add api to accept SNS notifications of S3 file added (#736)

### LINKED ISSUE #736: S3 Upload Notification Ingestion System
# EPIC: S3 Notification Ingestion System

## Overview
Implement a secure, reliable system to capture and process S3 object creation notifications for file tracking and integrity validation in the HCA Atlas Tracker.

### **Event Capture**
- Accept S3 object creation notifications via AWS SNS
- Support SNS message format with embedded S3 event data
- Handle individual S3 object events (not batch processing)

### **Authentication & Security**
- Verify SNS message signatures to prevent unauthorized requests
- Validate S3 event structure and required metadata
- Reject malformed or incomplete notifications

### **Data Persistence**
- Store file metadata in PostgreSQL database
- Capture: bucket, key, version_id, etag, size, SHA256 checksum
- Support file versioning with latest-version tracking
- Maintain data integrity constraints

### **Idempotency Requirements**
- Handle duplicate notifications gracefully without data corruption
- Detect and log ETag mismatches for same S3 object versions
- Ensure atomic operations for concurrent request handling
- Provide clear distinction between duplicates and new versions

### **Error Handling**
- Return appropriate HTTP status codes (400, 401, 500)
- Log operational events for monitoring and debugging
- Provide detailed error messages for troubleshooting
- Handle database transaction failures gracefully

### **Performance Requirements**
- Single atomic database operation per notification
- Avoid race conditions in concurrent processing
- Minimize database queries through efficient conflict resolution
- Support high-throughput S3 notification processing

### **Data Integrity**
- Validate SHA256 checksums from S3 object metadata
- Enforce strict format validation (64-character hex)
- Detect potential data corruption through ETag validation
- Maintain referential integrity in database operations

## Success Criteria
- ✅ Secure SNS authentication prevents unauthorized access
- ✅ Idempotent processing handles duplicate notifications safely
- ✅ File versioning tracks latest versions efficiently
- ✅ Comprehensive error handling and logging for production monitoring
- ✅ High performance with atomic database operations
- ✅ Complete test coverage for all scenarios

## Technical Notes
- Implementation uses Next.js API routes for HTTP endpoint
- PostgreSQL ON CONFLICT for atomic idempotency handling
- Comprehensive test suite covers authentication, validation, and edge cases
- Design decisions documented for future reference

### PR DESCRIPTION
**Pull request by NoopDog**
Closes #736 

This pull request introduces a comprehensive AWS resource authorization layer for S3 notification processing, along with detailed documentation for AWS resource configuration, entity relationships, and S3 notification design decisions. The main code addition is a utility for validating authorized SNS topics and S3 buckets using a JSON configuration, which enhances security and environment isolation. The accompanying documentation provides clear setup instructions, security rationale, and integration details for both developers and operations.

**AWS Resource Authorization and Configuration**

- Added `app/config/aws-resources.ts` with functions to parse and cache the `AWS_RESOURCE_CONFIG` environment variable, and to check if a given SNS topic or S3 bucket is authorized. This ensures only allowed AWS resources can trigger application logic and provides security logging for unauthorized access attempts.
- Added `docs/aws-resource-configuration.md` with setup instructions, configuration structure, environment-specific examples, security benefits, error responses, and configuration management details for the new resource authorization system.

**Entity Relationship and Data Model Documentation**

- Added `docs/entity-relationship-model.md` describing the hierarchical data model of the HCA Atlas Tracker, including detailed entity relationships, files table schema, processing workflow, error handling, and query patterns. This clarifies how files are integrated into the project’s entity model and enforces business rules at the schema level.

**S3 Notification System Design Decisions**

- Added `docs/s3-notification-design-decisions.md` outlining architectural, security, validation, idempotency, and error handling decisions for the S3 notification ingestion system. This document explains the rationale behind key design choices and summarizes implementation details and best practices.

### CODE REVIEW

File: scripts/migration-runner.ts [hunk 1]
```diff
@@ -4,14 +4,16 @@ import { getPoolConfig } from "../app/utils/pg-migrate-connect-config";
...
 const { Pool } = pg;
 
+const green = (text: string): string => `\x1b[32m${text}\x1b[0m`;
+
 const runMigrations = async (): Promise<void> => {
   const poolConfig = getPoolConfig();
   const pool = new Pool(poolConfig);
   const client = await pool.connect();
-  console.log`Connected to database: ${poolConfig.database}`;
+  console.log(green(`Connected to database: ${poolConfig.database}`));
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply scripts/migration-runner.ts:13]: Copilot seems to see the code incorrectly. The solution already exists in this  file

File: scripts/migration-runner.ts [hunk 2]
```diff
@@ -4,14 +4,16 @@ import { getPoolConfig } from "../app/utils/pg-migrate-connect-config";
...
 
   await client.query("CREATE SCHEMA IF NOT EXISTS hat");
-  console.log`Schema hat created successfully`;
+  console.log(green(`Schema hat created successfully`));
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply scripts/migration-runner.ts:16]: Copilot seems to see the code incorrectly. The solution already exists in this  file

File: migrations/1754790000000_files-table.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+ * Version Management: is_latest
+ * File Classification: file_type
+ * Foreign Key Relationships: source_study_id, atlas_id
+ * S3 Event Context: event_info
+ * Status & Processing: status
+ *
+ * @param pgm - Migration builder instance
+ */
+export const up = (pgm: MigrationBuilder): void => {
+  pgm.createTable(
```
[the remark this answers is not in the document]
Comment (NoopDog) [reply migrations/1754790000000_files-table.ts:22]: Yes, JSONB is empty in this instance. Also, there are a lot of indexes. Will take a look.

File: pages/api/files/s3-notification.ts
[the remark this answers is not in the document]
Comment (NoopDog) [reply pages/api/files/s3-notification.ts:1]: Excellent feedback from your colleague! This is exactly the kind of architectural guidance that leads to maintainable codebases. Let me examine the existing API patterns first to understand the established structure. ;)

File: docs/api-creation-guide.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+```typescript
+// Only support POST
+export default handler(
+  method(METHOD.POST), // Only POST is allowed
+  async (req, res) => {
+    // Handler logic
+  }
+);
+
+// Support multiple methods
```
Comment (hunterckx) [inline docs/api-creation-guide.md:225]: This should be done with the `handleByMethod` function; see e.g. `pages/api/atlases/[atlasId].ts`

File: docs/api-creation-guide.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+export enum METHOD {
+  DELETE = "DELETE",
+  GET = "GET",
+  PATCH = "PATCH",
+  POST = "POST",
+  PUT = "PUT",
+}
+```
+
+## 8. Custom Error Status Code Mapping
```
Comment (hunterckx) [inline docs/api-creation-guide.md:251]: This is not based on any existing convention of ours and maybe contradicts the strategy established elsewhere in the PR of extending the error types from api-handler.ts?

**Review by hunterckx — CHANGES_REQUESTED** [review]

File: pages/api/sns.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  }
+}
+
+/**
+ * Validates the cryptographic signature of an SNS message
+ * @param message - The SNS message to validate
+ * @throws SNSSignatureValidationError if the SNS signature is invalid (401)
+ * @throws InvalidOperationError if the message is malformed (400)
+ */
+async function validateSNSSignature(message: SNSMessage): Promise<void> {
```
[the remark this answers is not in the document]
Comment (hunterckx) [reply pages/api/sns.ts:87]: `async` just makes the function automatically return a promise and allows it to use `await`; since all this function does is return a promise directly, `async` can be removed without affecting anything. I would probably change it (at least when working in TypeScript where the type is a more reliable way of telling that a function returns a promise) in order to avoid implying that there's something in the function that requires `async`, but it's mostly a stylistic decision and not a big deal either way

(Side note: the AI might be conflating the `async` keyword with the concept of asynchrony? As long as the function returns a promise it's still asynchronous and gets called in the same way, even if it's not defined with the `async` keyword)
